ligo/stablecoin/actions.ligo

```diff
@@ -0,0 +1,288 @@
+(* SPDX-FileCopyrightText: 2020 tqtezos
+ * SPDX-License-Identifier: MIT *)
+
+(*
+ * Root entrypoints for stablecoin smart-contract
+ *)
+
+#include "operator.ligo"
+
+(* ------------------------------------------------------------- *)
+
+(*
+ * Helpers and validation
+ *)
+
+(*
+ * This function fails if provided token_id is not equal to
+ * default one, restricting all operations to be monolitic
+ *)
+function validate_token_type
+  ( const token_id : token_id
+  ) : unit is
+    if token_id =/= default_token_id
+    then failwith ("This operation does not support multiple token types")
+    else unit
+
+(*
+ * Same as above but for a list of token ids
+ *)
+function validate_token_types
+  ( const token_ids : list (token_id)
+  ) : unit is List.fold
+    ( function
+        ( const u        : unit
+        ; const token_id : token_id
+        ) : unit is validate_token_type (token_id)
+    , token_ids
+    , unit
+    )
+
+(*
+ * Increase the balance of an address associated with it
+ * and present in ledger. If not, create a new address with
+ * that balance given.
+ *)
+function credit_to
+  ( const parameter : mint_params
+  ; const store     : storage
+  ) : storage is block
+{ var updated_balance : nat := 0n
+; case store.ledger[parameter.to_] of
+    Some (ledger_balance) -> updated_balance := ledger_balance
+  | None -> skip
+  end
+; updated_balance := updated_balance + parameter.amount
+; const updated_ledger : ledger =
+    Big_map.update(parameter.to_, Some (updated_balance), store.ledger)
+; const ups : storage = store with record [ ledger = updated_ledger ]
+} with ups
+
+(*
+ * Decrease the balance of an address that is present in
+ * ledger, fail if it's not or given address has not enough
+ * tokens to be burn from.
+ *)
+function debit_from
+  ( const parameter : burn_params
+  ; const store     : storage
+  ) : storage is block
+{ var updated_ledger : ledger := store.ledger
+; case store.ledger[parameter.from_] of
+    Some (ledger_balance) -> if parameter.amount > ledger_balance
+      then failwith ("Given address has not enough balance to debit from")
```

Comment:
Recently error specification was added to FA2. Since it was added after we started our work, I think we can update our code later, but ideally we should follow it and fail with `INSUFFICIENT_BALANCE` here (and with `TOKEN_UNDEFINED` in `validate_token_type`).